src/tools/crawler/crawler.rs

```diff
@@ -0,0 +1,107 @@
+use std::{net::SocketAddr, str::FromStr, sync::Arc};
+
+use pea2pea::protocols::{Handshake, Reading, Writing};
+use tokio::{
+    sync::mpsc::{channel, Receiver},
+    time::{timeout, Instant},
+};
+use tracing::{info, trace, warn};
+use ziggurat_xrpl::{
+    protocol::{
+        codecs::message::{BinaryMessage, Payload},
+        proto::TmEndpoints,
+    },
+    setup::constants::CONNECTION_TIMEOUT,
+    tools::inner_node::InnerNode,
+};
+
+use crate::network::KnownNetwork;
+
+pub(super) struct Crawler {
+    node: InnerNode,
+    known_network: Arc<KnownNetwork>,
+    receiver: Receiver<(SocketAddr, BinaryMessage)>,
+}
+
+impl Crawler {
+    pub(super) async fn new() -> Self {
+        let (sender, receiver) = channel(1024);
+        let node = InnerNode::new(&Default::default(), sender).await;
+        node.enable_handshake().await;
+        node.enable_reading().await;
+        node.enable_writing().await;
+        let known_network = Arc::new(Default::default());
+        Self {
+            node,
+            known_network,
+            receiver,
+        }
+    }
+
+    pub(super) async fn start_processing(&mut self) {
+        while let Some((addr, message)) = self.receiver.recv().await {
+            if let Payload::TmEndpoints(endpoints) = message.payload {
+                self.process_endpoints_message(addr, endpoints).await;
+            }
+        }
+    }
+
+    async fn process_endpoints_message(&mut self, addr: SocketAddr, endpoints: TmEndpoints) {
+        info!(
+            "Received endpoints from {}: {:?}",
+            addr, endpoints.endpoints_v2
+        );
+        for endpoint in endpoints.endpoints_v2 {
+            match SocketAddr::from_str(&endpoint.endpoint) {
+                Ok(peer) => {
+                    self.process_peer(addr, endpoint.hops, peer).await;
+                }
+                Err(e) => {
+                    warn!("invalid address from {}: {}", addr, e);
+                }
+            }
+        }
+    }
+
+    async fn process_peer(&mut self, addr: SocketAddr, hops: u32, peer: SocketAddr) {
```

Comment:
```suggestion
    async fn process_peer(&mut self, addr: SocketAddr, hops: u32, addr: SocketAddr) {
```

Reply:
To be honest I prefer this name. With this change we would have two `addr` params.

Reply:
Sure!
I know that I was confused by the `peer` name when I saw it, so I had to check what it meant - basically, it took me more time to understand the code, and this suggestion would probably help me to understand it faster, but it what works best for me doesn't necessarily work for others 😅

Reply:
Let's leave it as peer for now and see if anything better comes up in the future.